RUNNERNEXT function on standalone

pi-r-p opened the pull request:
This new function allow to reschedule a runner from the runner itself.

To test, generate a token with this capability:
```
  'attributes' {
    '.cap:runner.reschedule.min.period' '5000'
  }
```

Use it with CAPADD inside your runner. From the runner, you can schedule the next iteration to "start of the script" + the period you want:
10 s RUNNERNEXT

It will fail if the requested period is less than the one defined in capability.

The function have no effect on a standalone with remote runner execution, or on a distributed instance.

hbs commented on warp10/src/main/java/io/warp10/script/functions/RUNNERNEXT.java:
```diff
@@ -0,0 +1,68 @@
+    // convert to milliseconds    
+    Long p = ((Long) o) / Constants.TIME_UNITS_PER_MS;
+    if (p < minPeriod) {
+      throw new WarpScriptException(getName() + " cannot set period below " + minPeriod + "ms defined in " + WarpScriptStack.CAPNAME_RUNNER_RESCHEDULE_MIN_PERIOD + " capability.");
```
add a space before 'ms'

pi-r-p commented:
doc OK. https://warp10.io/doc/RUNNERNEXT

pi-r-p commented:
- Fixed overflow if `MAXLONG RUNNERNEXT`
- introduced `RUNNERPROG` to schedule at a given time: 
```
"2022-04-09T21:48:00.0Z" TOTIMESTAMP RUNNERPROG
```

Both combined can allow synchronization on absolute time easily:
```
"2022-04-09T21:48:00.0Z" TOTIMESTAMP 'startsync' STORE
$startsync NOW >
<%
  MAXLONG RUNNERNEXT
  $startsync RUNNERPROG
%>
<%
  1 h RUNNERNEXT
%>
 IFTE
```

hbs commented:
Should rename the functions RUNNERIN and RUNNERAT

On Sat, Apr 9, 2022, 23:58 Pierre Papin ***@***.***> wrote:

>
>    - Fixed overflow if MAXLONG RUNNERNEXT
>    - introduced RUNNERPROG to schedule at a given time:
>
> "2022-04-09T21:48:00.0Z" TOTIMESTAMP RUNNERPROG
>
> Both combined can allow synchronization on absolute time easily:
>
> "2022-04-09T21:48:00.0Z" TOTIMESTAMP 'startsync' STORE
> $startsync NOW >
> <%
>   MAXLONG RUNNERNEXT
>   $startsync RUNNERPROG
> %>
> <%
>   1 h RUNNERNEXT
> %>
>  IFTE
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/senx/warp10-platform/pull/1088#issuecomment-1094130701>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABZYMRGXOBZLHCBEA7BSE3VEH4QDANCNFSM5STHAZKA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

hbs commented on warp10/src/main/java/io/warp10/script/WarpScriptStack.java:
```diff
@@ -302,6 +302,21 @@ public static enum Signal {
+  public static final String ATTRIBUTE_RUNNER_RESCHEDULE_PERIOD = "runner.reschedule.period";
+
+  /**
+   * Stack attribute for runner rescheduling period in millisecond
```
Comment seems off.

hbs commented on warp10/src/main/java/io/warp10/script/WarpScriptStack.java:
```diff
@@ -302,6 +302,21 @@ public static enum Signal {
   public static final String ATTRIBUTE_TIMEBOXED = "stack.timeboxed";
 
+  /**
+   * Cap name for runner rescheduling minimum period
```
Add expected unit.

hbs commented on warp10/src/main/java/io/warp10/script/functions/RUNNERIN.java:
```diff
@@ -0,0 +1,70 @@
+//
+//   Copyright 2018-2022  SenX S.A.S.
```
2022

hbs commented on warp10/src/main/java/io/warp10/script/functions/RUNNERAT.java:
```diff
@@ -0,0 +1,71 @@
+//
+//   Copyright 2018-2022  SenX S.A.S.
```
2022

hbs commented on warp10/src/main/java/io/warp10/script/WarpScriptLib.java:
```diff
@@ -678,6 +678,8 @@ public class WarpScriptLib {
   public static final String AESWRAP = "AESWRAP";
   public static final String AESUNWRAP = "AESUNWRAP";
   public static final String RUNNERNONCE = "RUNNERNONCE";
+  public static final String RUNNERAT = "RUNNERPROG";
```
Names of functions were not changed.

hbs commented on warp10/src/main/java/io/warp10/standalone/StandaloneScriptRunner.java:
```diff
@@ -224,19 +238,44 @@ public void run() {
             stack.execMulti(mc2WithReplacement.toString());
+
+            // Did the user asked to reschedule script to another period with RUNNERNEXT ?
+            if (stack.getAttribute(WarpScriptStack.ATTRIBUTE_RUNNER_RESCHEDULE_PERIOD) != null) {
```
Change check to 'instanceof Long'

hbs commented on warp10/src/main/java/io/warp10/standalone/StandaloneScriptRunner.java:
```diff
@@ -224,19 +238,44 @@ public void run() {
+              periodicityForNextRun = (Long) stack.getAttribute(WarpScriptStack.ATTRIBUTE_RUNNER_RESCHEDULE_PERIOD);
+            }
+            // Did the user asked to reschedule script to another absolute time with RUNNERPROG ? (absolute milliseconds)
+            if (stack.getAttribute(WarpScriptStack.ATTRIBUTE_RUNNER_RESCHEDULE_TIMESTAMP) != null) {
```
Ditto

hbs commented on warp10/src/main/java/io/warp10/standalone/StandaloneScriptRunner.java:
```diff
@@ -224,19 +238,44 @@ public void run() {
+            }
+
+            // progForNextRun is absolute time, not jvm relative time
+            long jvmStartTimeOffset = System.currentTimeMillis() - (System.nanoTime() / 1000000L);
```
Since System.nanoTime is unrelated to currentTimeMillis, use exact arithmetic to avoir overflows. Explain in comment what this code is doing, it is not that clear.

pi-r-p replied:
Both are positive, both are long. How can this operation lead to an overflow ?
Math.xxxExact throw exceptions, I don't want to crash on overflow, but managed it to be able to use RUNNERIN and RUNNERAT with MAXLONG

pi-r-p replied:
Moved timeMillis to NanoTime conversion into TimeSource, to be consistent.

hbs commented on warp10/src/main/java/io/warp10/standalone/StandaloneScriptRunner.java:
```diff
@@ -115,11 +122,17 @@ public void run() {
           ByteArrayOutputStream baos = new ByteArrayOutputStream();
 
+          long periodicityForNextRun = periodicity;
+          long progForNextRun = Long.MAX_VALUE;
```
Rename prod* variables into something clearer. This is not obvious since the name of the function RUNNERPROG was changed.

pi-r-p commented:
ready for new review.

hbs commented on warp10/src/main/java/io/warp10/script/functions/RUNNERAT.java:
```diff
@@ -0,0 +1,71 @@
+      throw new WarpScriptException(getName() + " expects a LONG timestamp as parameter.");
+    }
+    // convert to milliseconds    
+    Long progTs = ((Long) o) / Constants.TIME_UNITS_PER_MS;
```
use primitive type long

hbs commented on warp10/src/main/java/io/warp10/script/functions/RUNNERAT.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    // rescheduling capability is defined in milliseconds (as runners subdirectories)
+    Long minPeriod = Long.MAX_VALUE;
```
use primitive type long

hbs commented on warp10/src/main/java/io/warp10/script/functions/RUNNERIN.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    // rescheduling capability is defined in milliseconds (as runners subdirectories)
+    Long minPeriod = Long.MAX_VALUE;
```
use primitive type long

hbs replied:
change to long primitive type

hbs commented on warp10/src/main/java/io/warp10/script/functions/RUNNERIN.java:
```diff
@@ -0,0 +1,70 @@
+      throw new WarpScriptException(getName() + " expects a LONG period as parameter.");
+    }
+    // convert to milliseconds    
+    Long p = ((Long) o) / Constants.TIME_UNITS_PER_MS;
```
use primitive type long

hbs commented on warp10/src/main/java/io/warp10/standalone/StandaloneScriptRunner.java:
```diff
@@ -64,6 +65,12 @@ public class StandaloneScriptRunner extends ScriptRunner {
+   * runContexts < script path , hashmap > can store context objects for next runner iteration.
+   * Currently stores only runner.execution.counter, but can be extended later.
+   */
+  final Map<String, HashMap> runContexts = new ConcurrentHashMap<String, HashMap>();
```
the runContexts map is never purged, which will lead to leaking in the case of long running Warp 10 instances with an evolving list of runners.

The map needs to be cleaned periodically, maybe after scheduling?

pi-r-p replied:
Done. There was also a leak for nextrun hashmap, I fixed the problem.

pi-r-p commented:
ready for new review.

hbs commented on warp10/src/main/java/io/warp10/continuum/store/Constants.java:
```diff
@@ -120,7 +120,8 @@ public class Constants {
   public static final String RUNNER_SCHEDULEDAT = "runner.scheduledat";
   public static final String RUNNER_NONCE = "runner.nonce";
-
+  public static final String RUNNER_CONTEXT_EXEC_COUNT = "runner.execution.counter";
```
rename to runner.execution.count
